### PR TITLE
Alch

### PR DESCRIPTION
**Pull request by jamosaur**
### Description:

-  Adds alch command

### Changes:

-   [x] I have tested all my changes thoroughly.

### CODE REVIEW

File: src/commands/Minion/alch.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+			'Staff of fire',
+			'Fire battlestaff',
+			'Mystic fire staff',
+			'Lava battlestaff',
+			'Mystic lava staff',
+			'Steam battlestaff',
+			'Mystic steam staff',
+			'Smoke battlestaff',
+			'Mystic smoke staff',
+			'Tome of fire (empty)',
```
Comment (Alexsuperfly) [inline src/commands/Minion/alch.ts:50]: this doesnt give fire runes afaik

File: src/commands/Minion/alch.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+				}
+			}
+		}
+
+		const alchValue = quantity * highalch;
+
+		const consumedItems = resolveNameBank({
+			...(fireRuneCost > 0 ? { 'Fire rune': fireRuneCost } : {}),
+			'Nature rune': natureRuneCost,
+			[itemName]: quantity
```
Comment (Alexsuperfly) [inline src/commands/Minion/alch.ts:81]:
```suggestion
			[name]: quantity
```

getOSItem cleans the name passed to it so it could technically not be the true name of the item so use the returned OSItems name

File: src/commands/Minion/alch.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+		for (const runeProvider of unlimitedFireRuneProviders) {
+			if (
+				hasItemEquipped(runeProvider, msg.author.settings.get(UserSettings.Gear.Skilling))
+			) {
+				fireRuneCost = 0;
+				break;
+			}
+		}
+
+		const { id: bryophytasStaffId } = getOSItem("Bryophyta's staff");
```
Comment (Alexsuperfly) [inline src/commands/Minion/alch.ts:63]: you can just declare this id as a const at the top of file

File: src/commands/Minion/alch.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+				return alchMessage.edit(`Cancelling alch of ${quantity}x ${name}`);
+			}
+		}
+
+		await msg.author.settings.update(
+			UserSettings.Bank,
+			removeBankFromBank(userBank, consumedItems)
+		);
+
+		await msg.author.settings.update(
```
Comment (Alexsuperfly) [inline src/commands/Minion/alch.ts:121]: use the `addGP` method for this

File: src/commands/Minion/alch.ts [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+		}
+
+		const { id: bryophytasStaffId } = getOSItem("Bryophyta's staff");
+
+		if (
+			hasItemEquipped(bryophytasStaffId, msg.author.settings.get(UserSettings.Gear.Skilling))
+		) {
+			for (let i = 0; i < quantity; i++) {
+				if (roll(15)) {
+					natureRuneCost--;
```
Comment (Alexsuperfly) [inline src/commands/Minion/alch.ts:70]: doing this first before checking if they have the required runes could lead to a case where the user can try over and over again for a "good" roll to save the most nature runes.  do the rune saving after checking they have all the required items

File: src/tasks/minions/alchingActivity.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+		itemName,
+		quantity,
+		channelID,
+		alchValue,
+		userID,
+		savedRunes
+	}: AlchingActivityTaskOptions) {
+		const user = await this.client.users.fetch(userID);
+
+		await user.settings.update(UserSettings.GP, user.settings.get(UserSettings.GP) + alchValue);
```
Comment (Alexsuperfly) [inline src/tasks/minions/alchingActivity.ts:18]: use the `user.addGP` method here

File: src/tasks/minions/alchingActivity.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+		// If there are any saved runes from bryophyta's staff, return them
+		if (savedRunes) {
+			const returnedRunes = resolveNameBank({
+				'Nature rune': savedRunes
+			});
+			const userBank = await user.settings.get(UserSettings.Bank);
+			const newBank = await addBanks([userBank, returnedRunes]);
+			await user.settings.update(UserSettings.Bank, newBank);
```
Comment (Alexsuperfly) [inline src/tasks/minions/alchingActivity.ts:27]: use the `user.addItemsToBank` method here

File: src/tasks/minions/alchingActivity.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+			} has finished alching ${quantity}x ${itemName}! ${alchValue.toLocaleString()}gp (${toKMB(
+				alchValue
+			)}) has been added to your bank`,
+			savedRunes > 0 ? `Your Bryophyta's staff saved you ${savedRunes} Nature runes` : ''
+		];
+
+		this.client.queuePromise(() => {
+			channel.send(responses.join('. '));
+		});
+	}
```
Comment (Alexsuperfly) [inline src/tasks/minions/alchingActivity.ts:45]: traditionally any activity that takes time and can cause someone to not be able to do all of what they want to in one trip, the option to resend the trip that was just completed by responding with "yes"/"y" is something they are given.  take a look at other activities for examples

File: src/lib/constants.ts [hunk 9]
```diff
@@ -115,7 +115,8 @@ export const enum Tasks {
 	MonsterKillingTicker = 'monsterKillingTicker',
 	ClueTicker = 'clueTicker',
 	SkillingTicker = 'skillingTicker',
-	MinigameTicker = 'minigameTicker'
+	MinigameTicker = 'minigameTicker',
+	AlchingActivity = 'alchingActivity'
```
Comment (Alexsuperfly) [inline src/lib/constants.ts:119]: place this above the tickers and with the other activities

File: src/commands/Minion/alch.ts [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+		);
+
+		const userBank = await msg.author.settings.get(UserSettings.Bank);
+
+		if (!bankHasAllItemsFromBank(userBank, consumedItems)) {
+			throw `You don't have the required items, you need ${consumedItemsString}`;
+		}
+
+		if (
+			hasItemEquipped(bryophytasStaffId, msg.author.settings.get(UserSettings.Gear.Skilling))
```
Comment (Alexsuperfly) [inline src/commands/Minion/alch.ts:108]: since saved runes isnt used in this file, how about calculating it in the activity

File: src/commands/Minion/alch.ts [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+		);
+
+		const data: AlchingActivityTaskOptions = {
+			itemName: name,
+			userID: msg.author.id,
+			channelID: msg.channel.id,
+			quantity,
+			duration,
+			alchValue,
+			bryophytasStaffEquipped: hasItemEquipped(
```
Comment (Alexsuperfly) [inline src/commands/Minion/alch.ts:137]: i dont think you need to snapshot the gear when sending out because minion gear should be locked while in an activity afaik

**Comment by Alexsuperfly** [comment]
you need to place a section in the [user extendables minion file ](https://github.com/oldschoolgg/oldschoolbot/blob/dev/src/extendables/User/Minion.ts) to output a string saying a minion is alching while the alching activity is active

**Comment by jamosaur** [comment]
> you need to place a section in the [user extendables minion file ](https://github.com/oldschoolgg/oldschoolbot/blob/dev/src/extendables/User/Minion.ts) to output a string saying a minion is alching while the alching activity is active

Added 😄

File: src/commands/Minion/alch.ts [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+		const maxCasts = Math.min(Math.floor(msg.author.maxTripLength / timePerAlch), numOfItems);
+
+		if (quantity * timePerAlch > msg.author.maxTripLength) {
+			throw `The max number of alchs you can do is ${maxCasts}!`;
+		}
+
+		const duration = quantity * timePerAlch;
+		let fireRuneCost = quantity * 5;
+
+		const unlimitedFireRuneProviders = resolveItems([
```
Comment (Alexsuperfly) [inline src/commands/Minion/alch.ts:61]: define this outside of the `run` function so it doesnt have to be remade every execution

File: src/extendables/User/Minion.ts [hunk 13]
```diff
@@ -4,20 +4,21 @@ import { User } from 'discord.js';
 import getActivityOfUser from '../../lib/util/getActivityOfUser';
 import { formatDuration } from '../../util';
 import {
-	MonsterActivityTaskOptions,
-	ClueActivityTaskOptions,
-	CraftingActivityTaskOptions,
 	AgilityActivityTaskOptions,
+	AlchingActivityTaskOptions,
```
Comment (Alexsuperfly) [inline src/extendables/User/Minion.ts:8]: revert the import reordering change, its unnecessary diff

File: src/extendables/User/Minion.ts [hunk 14]
```diff
@@ -4,20 +4,21 @@ import { User } from 'discord.js';
...
+	MonsterActivityTaskOptions,
+	OfferingActivityTaskOptions,
 	SmeltingActivityTaskOptions,
 	SmithingActivityTaskOptions,
-	FiremakingActivityTaskOptions,
-	WoodcuttingActivityTaskOptions,
-	OfferingActivityTaskOptions,
-	BuryingActivityTaskOptions,
-	FletchingActivityTaskOptions
+	WoodcuttingActivityTaskOptions
```
Comment (gc) [inline src/extendables/User/Minion.ts:21]: please undo whatever changes you made to the ordering of these. a reorganization like this is better suited for a standalone PR, and isnt something that should go in a PR like this, it makes reviewing harder and adds lots of diffs.

File: src/lib/types/minions.ts [hunk 15]
```diff
@@ -137,22 +144,23 @@ export type TickerTaskData =
...
-	| Tasks.WoodcuttingActivity
-	| Tasks.RunecraftActivity
-	| Tasks.FiremakingActivity
-	| Tasks.QuestingActivity
-	| Tasks.BuryingActivity
-	| Tasks.OfferingActivity
-	| Tasks.FightCavesActivity
-	| Tasks.FletchingActivity
-	| Tasks.WintertodtActivity;
+	| Tasks.WintertodtActivity
```
Comment (gc) [inline src/lib/types/minions.ts:165]: same as said above

File: src/tasks/minions/alchingActivity.ts [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+import { Time } from '../../lib/constants';
+import hasItemEquipped from '../../lib/gear/functions/hasItemEquipped';
+import { UserSettings } from '../../lib/settings/types/UserSettings';
+import getOSItem from '../../lib/util/getOSItem';
+
+const { id: bryophytasStaffId } = getOSItem("Bryophyta's staff");
+
+export default class extends Task {
+	async run({ itemName, quantity, channelID, alchValue, userID }: AlchingActivityTaskOptions) {
+		let savedRunes = 0;
```
Comment (gc) [inline src/tasks/minions/alchingActivity.ts:16]: move the declaration of `savedRunes` below above the if statement

File: src/tasks/minions/alchingActivity.ts [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+import { channelIsSendable } from '../../lib/util/channelIsSendable';
+import { resolveNameBank, toKMB } from 'oldschooljs/dist/util';
+import { noOp, roll, saidYes } from '../../lib/util';
+import getUsersPerkTier from '../../lib/util/getUsersPerkTier';
+import { Time } from '../../lib/constants';
+import hasItemEquipped from '../../lib/gear/functions/hasItemEquipped';
+import { UserSettings } from '../../lib/settings/types/UserSettings';
+import getOSItem from '../../lib/util/getOSItem';
+
+const { id: bryophytasStaffId } = getOSItem("Bryophyta's staff");
```
Comment (gc) [inline src/tasks/minions/alchingActivity.ts:12]: `const bryophytasStaffId  = itemID("Bryophyta's staff")`
the itemID fn is made for this

File: src/commands/Testing/spawn.ts [hunk 18]
```diff
@@ -24,20 +27,27 @@ export default class extends BotCommand {
 			return;
 		}
 
-		const qty = typeof qtyOrName === 'number' ? qtyOrName : 1;
-		const itemName =
-			typeof qtyOrName === 'string' ? qtyOrName : itemNamePossible ?? 'Bunny feet';
+		if (msg.flagArgs.all) {
+			const items: Bank = {};
+			item.forEach(i => (items[i.id] = qty));
```
Comment (gc) [inline src/commands/Testing/spawn.ts:32]: use a `for of` loop for this

File: src/commands/Testing/spawn.ts [hunk 19]
```diff
@@ -1,20 +1,23 @@
...
-			usage: '[number:int{1,100}] <itemname:...string>',
+			usage: '[qty:integer{1,1000000}] (item:...item)',
 			usageDelim: ' ',
 			oneAtTime: true
 		});
 		this.enabled = !this.client.production;
 	}
 
-	async run(msg: KlasaMessage, [qtyOrName, itemNamePossible]: [number, number | undefined]) {
+	async run(msg: KlasaMessage, [qty = 1, item]: [number, Item[]]) {
```
Comment (gc) [inline src/commands/Testing/spawn.ts:20]: if `item` is always an array of items, it should probably be called `items` or `itemArray`

File: src/commands/Minion/alch.ts [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+			usage: '[quantity:int{1}] (item:...item)',
+			usageDelim: ' '
+		});
+	}
+
+	async run(msg: KlasaMessage, [quantity = null, item]: [number | null, Item[]]) {
+		const userBank = msg.author.settings.get(UserSettings.Bank);
+		const osItem = item.find(i => userBank[i.id] && i.highalch);
+		if (!osItem) {
+			throw `'You don't have any of this item to alch'`;
```
Comment (gc) [inline src/commands/Minion/alch.ts:53]: add a fullstop to the end

File: src/tasks/minions/alchingActivity.ts [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+			for (let i = 0; i < quantity; i++) {
+				if (roll(15)) savedRunes++;
+			}
+
+			if (savedRunes > 0) {
+				const returnedRunes = resolveNameBank({
+					'Nature rune': savedRunes
+				});
+
+				user.addItemsToBank(returnedRunes);
```
Comment (gc) [inline src/tasks/minions/alchingActivity.ts:32]: `await` this

File: src/commands/Minion/alch.ts [hunk 22]
```diff
@@ -0,0 +1,147 @@
...
+			duration,
+			alchValue,
+			type: Activity.Alching,
+			id: rand(1, 1_000_000),
+			finishDate: Date.now() + duration
+		};
+
+		await addSubTaskToActivityTask(this.client, Tasks.SkillingTicker, data);
+
+		await msg.author.incrementMinionDailyDuration(duration);
```
Comment (gc) [inline src/commands/Minion/alch.ts:138]: move this line to the alching task, e.g. `await user.incrementMinionDailyDuration(duration);` - directly below where the user var is

File: src/commands/Testing/setxp.ts [hunk 23]
```diff
@@ -0,0 +1,25 @@
+import { KlasaMessage, CommandStore } from 'klasa';
+import { BotCommand } from '../../lib/BotCommand';
+
+export default class extends BotCommand {
+	public constructor(store: CommandStore, file: string[], directory: string) {
+		super(store, file, directory, {
+			usage: '<amount:int{1,200000000}> <skillName:string>',
```
Comment (gc) [inline src/commands/Testing/setxp.ts:7]: this needs protection against wrong skill names being inputted, in some way. Maybe check that `Skills.some(skill => skill.id === skillName.toLowerCase())` or something like that

File: src/tasks/minions/alchingActivity.ts (hunk 16 above)
Comment (jamosaur) [reply src/tasks/minions/alchingActivity.ts:16]: Moving this below the if breaks things because it's required by the responses

File: src/commands/Minion/alch.ts [hunk 24]
```diff
@@ -0,0 +1,146 @@
...
+	async run(msg: KlasaMessage, [quantity = null, item]: [number | null, Item[]]) {
+		const userBank = msg.author.settings.get(UserSettings.Bank);
+		const osItem = item.find(i => userBank[i.id] && i.highalch);
+		if (!osItem) {
+			throw `'You don't have any of this item to alch.'`;
+		}
+		// 5 tick action
+		const timePerAlch = Time.Second * 3;
+
+		if (!quantity) {
```
Comment (Alexsuperfly) [inline src/commands/Minion/alch.ts:58]: move this below the maxCasts calc and set quantity to maxCasts, otherwise this could automatically set quant to something greater than you can do
